boa/src/builtins/array/mod.rs

```diff
@@ -461,30 +488,68 @@ impl Array {
     /// [spec]: https://tc39.es/ecma262/#sec-array.prototype.concat
     /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat
     pub(crate) fn concat(this: &Value, args: &[Value], context: &mut Context) -> Result<Value> {
-        if args.is_empty() {
-            // If concat is called with no arguments, it returns the original array
-            return Ok(this.clone());
-        }
-
-        // Make a new array (using this object as the prototype basis for the new
-        // one)
-        let mut new_values: Vec<Value> = Vec::new();
-
-        let this_length = this.get_field("length", context)?.to_length(context)?;
-        for n in 0..this_length {
-            new_values.push(this.get_field(n, context)?);
-        }
-
-        for concat_array in args {
-            let concat_length = concat_array
-                .get_field("length", context)?
-                .to_length(context)?;
-            for n in 0..concat_length {
-                new_values.push(concat_array.get_field(n, context)?);
+        // 1. Let O be ? ToObject(this value).
+        let obj = this.to_object(context)?;
+        // 2. Let A be ? ArraySpeciesCreate(O, 0).
+        let arr = Self::array_species_create(&obj, 0, context)?;
+        // 3. Let n be 0.
+        let mut n = 0;
+        // 4. Prepend O to items.
+        let mut items = args
+            .iter()
+            .flat_map(|item_val| item_val.to_object(context))
+            .collect::<VecDeque<GcObject>>();
```

Comment:
@Razican @HalidOdat Was hoping to get your input on whether I should use a VecDeque, ran a benchmark against both types with no consistent result proving better or worse.